contracts/src/matching/HyperdriveMatchingEngineV2.sol

```diff
@@ -0,0 +1,659 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.24;
+
+import { ECDSA } from "lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+import { EIP712 } from "lib/openzeppelin-contracts/contracts/utils/cryptography/EIP712.sol";
+import { ERC20 } from "lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
+import { IERC1271 } from "lib/openzeppelin-contracts/contracts/interfaces/IERC1271.sol";
+import { ReentrancyGuard } from "lib/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol";
+import { SafeERC20 } from "lib/openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import { AssetId } from "../libraries/AssetId.sol";
+import { FixedPointMath } from "../libraries/FixedPointMath.sol";
+import { HYPERDRIVE_MATCHING_ENGINE_KIND, VERSION } from "../libraries/Constants.sol";
+import { HyperdriveMath } from "../libraries/HyperdriveMath.sol";
+import { IHyperdrive } from "../interfaces/IHyperdrive.sol";
+import { IHyperdriveMatchingEngineV2 } from "../interfaces/IHyperdriveMatchingEngineV2.sol";
+
+/// @author DELV
+/// @title HyperdriveMatchingEngine
+/// @notice A matching engine that processes order intents and settles trades on 
+///         the Hyperdrive AMM.
+/// @dev This version uses direct Hyperdrive mint/burn functions instead of flash 
+///      loans.
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+contract HyperdriveMatchingEngineV2 is 
+    IHyperdriveMatchingEngineV2,
+    ReentrancyGuard,
+    EIP712
+{
+    using FixedPointMath for uint256;
+    using SafeERC20 for ERC20;
+
+    /// @notice The EIP712 typehash of the OrderIntent struct.
+    bytes32 public constant ORDER_INTENT_TYPEHASH =
+        keccak256(
+            "OrderIntent(address trader,address counterparty,address hyperdrive,uint256 amount,uint256 slippageGuard,uint256 minVaultSharePrice,Options options,uint8 orderType,uint256 minMaturityTime,uint256 maxMaturityTime,uint256 expiry,bytes32 salt)"
+        );
+
+    /// @notice The EIP712 typehash of the Options struct
+    bytes32 public constant OPTIONS_TYPEHASH =
+        keccak256("Options(address destination,bool asBase)");
+
+    /// @notice The name of this matching engine
+    string public name;
+
+    /// @notice The kind of this matching engine
+    string public constant kind = HYPERDRIVE_MATCHING_ENGINE_KIND;
+
+    /// @notice The version of this matching engine
+    string public constant version = VERSION;
```

Comment:
Nit: Make sure you're ending these sentences with periods. I'm seeing this elsewhere too, so it would be good for you to do a pass and fix this. This may seem pedantic, but it is about maintaining consistency.

```suggestion
    /// @notice The EIP712 typehash of the Options struct.
    bytes32 public constant OPTIONS_TYPEHASH =
        keccak256("Options(address destination,bool asBase)");

    /// @notice The name of this matching engine.
    string public name;

    /// @notice The kind of this matching engine.
    string public constant kind = HYPERDRIVE_MATCHING_ENGINE_KIND;

    /// @notice The version of this matching engine.
    string public constant version = VERSION;
```